### ISSUE TITLE
Soot crashed in a simple Java source

### ISSUE THREAD

**Issue by viennadd** [Sable/soot#867]
Hi all,

Our code scanner has found a Null Pointer Exception for the following case:
```java
// Main.java
public class Main {
  public int numTrees(int n) {
    int[] dp = null;
    return dp[0];
  }
}
```

![image](https://user-images.githubusercontent.com/3396288/34866002-e610ca70-f7b5-11e7-9abf-28627da08589.png)

Reproduction:

Use soot
`java -cp <path-to-sootclasses-trunk-jar-with-dependencies.jar> soot.Main -cp .:${JAVA_HOME}/jre/lib/rt.jar -f jimple Main`

It should raise an exception:
```
java.lang.RuntimeException: Failed to apply jb to <Main: int numTrees(int)>
    at soot.asm.AsmMethodSource.getBody(AsmMethodSource.java:1911)
    at soot.SootMethod.getBodyFromMethodSource(SootMethod.java:126)
    at soot.SootMethod.retrieveActiveBody(SootMethod.java:385)
    at soot.PackManager$3.run(PackManager.java:1216)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
    at soot.jimple.toolkits.typing.fast.UseChecker.caseAssignStmt(UseChecker.java:362)
    at soot.jimple.internal.JAssignStmt.apply(JAssignStmt.java:238)
    at soot.jimple.toolkits.typing.fast.UseChecker.check(UseChecker.java:125)
    at soot.jimple.toolkits.typing.fast.TypeResolver.insertCasts(TypeResolver.java:426)
    at soot.jimple.toolkits.typing.fast.TypeResolver.minCasts(TypeResolver.java:437)
    at soot.jimple.toolkits.typing.fast.TypeResolver.inferTypes(TypeResolver.java:158)
    at soot.jimple.toolkits.typing.TypeAssigner.internalTransform(TypeAssigner.java:126)
    at soot.BodyTransformer.transform(BodyTransformer.java:54)
    at soot.Transform.apply(Transform.java:105)
    at soot.JimpleBodyPack.applyPhaseOptions(JimpleBodyPack.java:73)
    at soot.JimpleBodyPack.internalApply(JimpleBodyPack.java:105)
    at soot.Pack.apply(Pack.java:125)
    at soot.asm.AsmMethodSource.getBody(AsmMethodSource.java:1909)
    ... 6 more
```

Regards,
Sourcebrella Inc.

**Comment by richardxx** [comment]
@ericbodden Maybe you would like to have a look, :>

**Comment by Alexandre-Bartel** [comment]
Seems like the type resolver was not able to detect the type of the array in that case. I updated it in branch [fix-867](https://github.com/Sable/soot/commit/88a86f42d6540c3f536a6ed5715d204b8cf262de)   so that it also extract type information from return statements. It works fine now on the example. More testing is required to check that this fix does not break anything else before merging it to develop.

**Comment by richardxx** [comment]
Thanks for the quick fix. Hope you can let us know once your patch is merged. Thanks.

**Comment by jpstotz** [comment]
This seems to be the same bug as reported in #809.

**Comment by mbenz89** [comment]
I've merged the fix.
@jpstotz Could you check if it solves your issues as well?
@Alexandre-Bartel Thanks for the quick fix!

**Comment by mbenz89** [comment]
Fixed in #877
